### PR TITLE
[WIP] OpenStack UPI improvements

### PR DESCRIPTION
**Pull request by adduarte**
Collection of "improvements" to upi based on user feedback and bug fixes.

### CODE REVIEW

**Comment by adduarte** [comment]
Currently mtu is only supported on os_network module for ansible 2.9 and higher.

File: upi/openstack/network.yaml [hunk 1]
```diff
@@ -14,6 +14,7 @@
   - name: 'Create the cluster network'
     os_network:
       name: "{{ os_network }}"
+      mtu: "{{ os_network_mtu }}"
```
Comment (adduarte) [inline upi/openstack/network.yaml:17]: "mtu" is only supported on os_network module for ansible 2.9 or higher.

**Comment by pierreprinetti** [comment]
@adduarte 
Is Ansible 2.9 distributed in RHEL 8? If so, we might just update [the docs](https://github.com/openshift/installer/blob/master/docs/user/openstack/install_upi.md#rhel) to fetch the right Ansible version.

If not, maybe we can add a note to the docs saying like "if you want to set the MTU, make sure you have Ansible 2.9 and add this parameter to os_network"

WDYT?

**Comment by adduarte** [comment]
How does this MTU setting affect a deployment kuryr is being used? 
Do we need to set the mtu for kuryr as well? 



> @adduarte
> Is Ansible 2.9 distributed in RHEL 8? If so, we might just update [the docs](https://github.com/openshift/installer/blob/master/docs/user/openstack/install_upi.md#rhel) to fetch the right Ansible version.
> 
> If not, maybe we can add a note to the docs saying like "if you want to set the MTU, make sure you have Ansible 2.9 and add this parameter to os_network while creating the network"
> 
> WDYT?

I think perhaps we should bump the required version of ansible to 2.9 since this would go into master.

**Comment by pierreprinetti** [comment]
@adduarte 
> I think perhaps we should bump the required version of ansible to 2.9

The steps we describe in the UPI docs must work at least on RHEL8. So before we bump, we must be sure that RHEL8 ships with Ansible 2.9.

Another thing: I don't think we set an MTU explicitly in IPI. Do you think there's the case for different behaviours?

**Comment by dulek** [comment]
So I think it'll also need to get added to https://github.com/openshift/installer/blob/master/upi/openstack/network.yaml#L36 and we might still have a bug that we don't set the MTU at all when creating new networks for pods in new namespaces…

**Comment by adduarte** [comment]
> So I think it'll also need to get added to https://github.com/openshift/installer/blob/master/upi/openstack/network.yaml#L36 and we might still have a bug that we don't set the MTU at all when creating new networks for pods in new namespaces…

I think we should test this. because form what I can tell the coreos VMs pick up the mtu from the network created. ANd if kuryr picks up the mtu from the core os vms, then setting it in the network is good enough.

**Comment by adduarte** [comment]
> @adduarte
> 
> > I think perhaps we should bump the required version of ansible to 2.9
> 
> The steps we describe in the UPI docs must work at least on RHEL8. So before we bump, we must be sure that RHEL8 ships with Ansible 2.9.
> 
> Another thing: I don't think we set an MTU explicitly in IPI. Do you think there's the case for different behaviours?

Correct in IPI, we do not set it. We take the default value, which is configurable by cloud (openstack) administrator. 
(when you create a network in openstck if you don't specify the mtu, you get the default mtu value). 

The problem is that sometimes that value is incorrect, because of misconfigured hardware or just simple mistakesm and will cause problems down the line for everyone involved.
The origin of this pull request was an actual customer trying to deploy and running into that exact same issue. Their cloud provider had wrong mtu set, thus they decided to set the mtu themselves. 
It might be best to merge this with that commented out just to show where it would be set if necessary.

**Review by mandre — COMMENTED** [review]
:+1:  for recommending the latest available version of ansible [in the docs](https://github.com/openshift/installer/blob/master/docs/user/openstack/install_upi.md#rhel). Ansible 2.9 should be [available in RHEL8](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html#installing-ansible-on-rhel-centos-or-fedora) in the ansible-2.9-for-rhel-8-x86_64-rpms repo.

I'm less sure about adding a config for MTU in the UPI scripts as the user is supposed to adapt the scripts for his environment. We probably do not want to start adding options for particular use cases.
Perhaps this is something that should better be documented in the [troubleshooting](https://github.com/openshift/installer/blob/master/docs/user/openstack/troubleshooting.md) or [known-issues](https://github.com/openshift/installer/blob/master/docs/user/openstack/known-issues.md) page.

**Comment by pierreprinetti** [comment]
@adduarte 

> So if a user wants to do something which can't be done with IPI, like say I don't know, set the MTU :), then they have to turn to UPI

I was thinking about the case where the default Neutron MTU is set to a different value, e.g. jumbo frames. In that case, the IPI installation would inherit the MTU=9000 since it does not explicitly set a value. However with this change, the untouched UPI behaves differently because it sets MTU=1400.

I think we should keep the exact same behaviour in IPI and UPI, and not set a value. I see two solutions:

## Solution 1
Add the code, but let the playbooks default to use Neutron's default. You can set `os_network_mtu: 0` to pick the default.

```yaml
# inventory.yaml
      # a value different than zero will override Neutron's default setting
      os_network_mtu: 0

# network.yaml
  tasks:
  - name: 'Create the primary cluster network'
    os_network:
      name: "{{ os_network }}"
      mtu:  "{{ os_mtu }}"
```

## Solution 2

Do not change the playbooks, and only document the change.

e.g. "You can explicitly set the cluster's MTU by adding the `mtu` property in the `os_network` task in `network.yaml`".

**Comment by jeichler** [comment]
Have you considered using something like this?

```yaml
  tasks:
  - name: 'Create the primary cluster network'
    os_network:
      name: "{{ os_network }}"
      mtu:  "{{ os_mtu | default(omit) }}"
```
if no variable is defined, it will simply omit the parameter entirely (and then use the OSP default according to the module documentation).

**Comment by dulek** [comment]
> > So I think it'll also need to get added to https://github.com/openshift/installer/blob/master/upi/openstack/network.yaml#L36 and we might still have a bug that we don't set the MTU at all when creating new networks for pods in new namespaces…
> 
> So I don't think you need to set it here unless you want to. The cluster network and service network are independent of each other. you could use an MTU of 9000 in one and an MTU of 1500 on another (as long as the openstack cloud provider supports it). They won't affect each other. traffic is separate since they are separate tenant networks.
> It would be "nice" to have as well but should not affect each other.

Okay, agreed!

**Comment by dulek** [comment]
> > So I think it'll also need to get added to https://github.com/openshift/installer/blob/master/upi/openstack/network.yaml#L36 and we might still have a bug that we don't set the MTU at all when creating new networks for pods in new namespaces…
> 
> So I don't think you need to set it here unless you want to. The cluster network and service network are independent of each other. you could use an MTU of 9000 in one and an MTU of 1500 on another (as long as the openstack cloud provider supports it). They won't affect each other. traffic is separate since they are separate tenant networks.
> It would be "nice" to have as well but should not affect each other.

Okay, agreed!

**Comment by adduarte** [comment]
It seems too constricting to force the user to select an mtu. 
Setting the mtu is something that should be left to user to add. 
Adding it to a playbook constricts more than helps.

**Pull request closed without merging**